providers/azure.go

```diff
@@ -230,44 +231,84 @@ func getEmailFromJSON(json *simplejson.Json) (string, error) {
 		err = otherMailsErr
 	}
 
+	if err != nil || email == "" {
+		email, err = json.Get("userPrincipalName").String()
+		if err != nil {
+			logger.Errorf("unable to find userPrincipalName: %s", err)
+			return "", err
+		}
+	}
+
 	return email, err
 }
 
-// GetEmailAddress returns the Account email address
-func (p *AzureProvider) GetEmailAddress(ctx context.Context, s *sessions.SessionState) (string, error) {
-	var email string
-	var err error
+// full claims that can be emitted by AAD are
+// https://docs.microsoft.com/en-us/azure/active-directory/develop/id-tokens
+// only include email here since we only need email extraction
+type aadIDTokenClaims struct {
+	Email string `json:"email,omitempty"`
+}
 
-	if s.AccessToken == "" {
+// Valid is the function validating the claims,
+// required by jwt.Claims interface.
+// returning nil as we only use it to extract claims instead of validating them
+func (aadIDTokenClaims) Valid() error {
+	return nil
+}
+
+func getEmailFromIDToken(idToken string) (string, error) {
+	claims := &aadIDTokenClaims{}
+	parser := jwt.Parser{}
+
+	// ParseUnverified to extract claims without verifying signature
+	if _, _, err := parser.ParseUnverified(idToken, claims); err != nil {
+		return "", err
+	}
+	if claims.Email == "" {
+		return "", errors.New("missing email claim from id_token")
+	}
+	return claims.Email, nil
+}
+
+func getEmailFromProfileAPI(ctx context.Context, accessToken, profileURL string) (string, error) {
+	if accessToken == "" {
 		return "", errors.New("missing access token")
 	}
 
-	json, err := requests.New(p.ProfileURL.String()).
+	json, err := requests.New(profileURL).
 		WithContext(ctx).
-		WithHeaders(makeAzureHeader(s.AccessToken)).
+		WithHeaders(makeAzureHeader(accessToken)).
 		Do().
 		UnmarshalJSON()
 	if err != nil {
 		return "", err
 	}
 
-	email, err = getEmailFromJSON(json)
-	if err == nil && email != "" {
-		return email, err
-	}
+	return getEmailFromJSON(json)
+}
 
-	email, err = json.Get("userPrincipalName").String()
-	if err != nil {
-		logger.Errorf("failed making request %s", err)
-		return "", err
+// EnrichSessionState finds the email to enrich the session state
+func (p *AzureProvider) EnrichSessionState(ctx context.Context, s *sessions.SessionState) error {
+	var email string
+	var err error
+
+	if s.IDToken != "" {
+		email, err := getEmailFromIDToken(s.IDToken)
+		if err == nil {
+			s.Email = email
+			return nil
+		}
+		logger.Errorf("unable to find email from id_token: %s", err)
```

Comment:
It's idiomatic to use `%v` in a format string for error types in Go, please switch this and other logger calls to use `%v", err` at the end
```suggestion
		logger.Errorf("unable to find email from id_token: %v", err)
```

Reply:
happy to change. though, do you happen to know where this idiom is documented? @JoelSpeed